[FIX #4145] Yoda conditions cop

smakagon opened the pull request:
This cop checks for Yoda conditions where the two parts of an expression
are reversed from the typical order in a conditional statement. (As discussed in #4145)
    
    # bad
    99 = foo
    "bar" == foo
 
    # good
    foo == 99
    foo == "bar"

Drenmi commented:
I'm curious what made you choose `if` and `while` as entry points, rather than hooking into `send` and checking the method called. 🙂 It seems overly complex to roll up nested conditions to extract the inner ones.

The current approach will always miss standalone expressions, e.g.:

```ruby
def meaning_of_life?(n)
  42 == n
end
```

WDYT?

smakagon commented:
@Drenmi I knew that something was wrong with that idea! I rewrote code, now it's 50 lines of readable code instead of 85 lines of mess. Thanks for help! Let me know what do you think about this implementation.

bbatsov commented:
I think this cop should be extended to cover comparison operators as well.

bbatsov commented on CHANGELOG.md:
```diff
@@ -29,6 +29,7 @@
 * Add new `Rails/Blank` cop. ([@rrosenblum][])
 * Add new `Rails/Present` cop. ([@rrosenblum][])
 * [#4004](https://github.com/bbatsov/rubocop/issues/4004): Allow not treating comment lines as group separators in `Bundler/OrderedGems` cop. ([@konto-andrzeja][])
+* [#4145](https://github.com/bbatsov/rubocop/issues/4145): Add new `Style/YodaCondition` cop. ([@smakagon][])
```
Due to yesterday's release you'll have to move this to the new `master` section of the changelog.

bbatsov commented:
Btw, this should also get a matching rule in the Ruby style guide.

Drenmi commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,52 @@
+        private
+
+        def comparison_operator?(node)
+          return false unless node.children.count == 3
```
I don't see any value in having this check. 🙂

Drenmi commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,52 @@
+        def comparison_operator?(node)
+          return false unless node.children.count == 3
+
+          RuboCop::AST::Node::COMPARISON_OPERATORS.include?(node.children[1])
```
There's a `SendNode` extension. Use it for less obscure destructuring:

```ruby
node.children[1]
```

vs.

```ruby
node.method_name
```

Drenmi commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,52 @@
+        end
+
+        def yoda_condition?(node)
+          !%i(lvar ivar send const regexp begin).include?(
```
I would put this array in a constant with a meaningful name. 🙂

Drenmi commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,52 @@
+        def on_send(node)
+          return unless comparison_operator?(node)
+
+          register_offense(node) if yoda_condition?(node)
```
Code looks really good now!

I would probably reorganize the conditions a bit. Currently, the `#yoda_condition?` method doesn't actually check if something is a Yoda condition. I would wrap the check for comparison operator and whitelist types in that method and then do this in the callback:

```ruby
return unless yoda_condition?(node)
```

Drenmi commented:
Starting to look good @smakagon! 🙂

bbatsov commented on config/enabled.yml:
```diff
@@ -1004,6 +1004,10 @@ Style/WordArray:
   Enabled: true
 
+Style/YodaCondition:
+  Description: 'Use x == 5 instead of 5 == x.'
```
I'd actually word this in more generic way - "don't use constants as the first operand of a comparison" or something like this.

mikegee replied:
constants or literals

bbatsov commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,52 @@
+        end
+
+        def register_offense(node)
+          first_value = node.children.first.source
```
I general I dislike the use of `node.children` - either use destructuring or the node extensions.

betesh commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,52 @@
+  module Cop
+    module Style
+      # This cop checks for Yoda conditions where the two parts of an expression
+      # are reversed from the typical order in a conditional statement.
```
What makes one order more typical than another?  Consider re-wording this to explain that the goal is to enforce that the order match the way that humans (or maybe just English-speaker?) speak.

smakagon replied:
@betesh How would you suggest to change that description?

betesh replied:
I took a closer look at the wikipedia article linked to on #4145 to get some ideas, and I realized this description is copied-and-pasted from there.  So whatever wording we choose here should probably be "back-ported" to the wikipedia article.

So I'll start by quoting from the end of that paragraph:
```
Yoda, who spoke English in a non-standard syntax.
```

and from later in the article (in the Criticism section):

```
Critics of Yoda conditions see the lack of readability as a disadvantage that
outweighs the benefits described above. 
```

With that in mind, I propose the following:
```
This cop checks for Yoda conditions, i.e. comparison operations where
readability is reduced because the operands are not ordered the same
way as they would be ordered in spoken English.
```

But feel free to change "spoken English" to "spoken language" if to your knowledge this is true in all other relevant languages here as well.

smakagon replied:
@Drenmi @bbatsov what do you guys think?

bbatsov replied:
Sounds reasonable to me. Similar wording can be used in the style guide I guess.

betesh commented on manual/cops_style.md:
```diff
@@ -6028,6 +6028,28 @@ WordRegex | (?-mix:\A[\p{Word}\n\t]+\z)
+
+Enabled by default | Supports autocorrection
+--- | ---
+Enabled | No
```
Would it be that difficult to support auto-correction?  (I've never done it before, so I'm not implying it's easy.  Just asking...)

smakagon replied:
@betesh I've never done this before as well, so I don't know yet. But that would be good to add in later iterations.

rrosenblum replied:
I don't think auto-correction would be difficult to implement for this cop.

smakagon replied:
Will look into that, that would be my first implementation of autocorrection.

smakagon commented:
Do you think I should do something else on this PR or we're good with this code/implementation?

bbatsov commented:
> I think this cop should be extended to cover all comparison operators as well.

5 < x, 5 <= x, etc. Maybe only `===` should be ignored.

rrosenblum commented:
Should this cop be configured to support the opposite style? I know that the intention here is to produce grammatically correct code, however having worked with Java in the past, it is ingrained in me to always place the "constant" before the variable to avoid a `NullPointerException`. Also the error message is a bit nicer
```ruby
[1] pry(main)> 5 < nil
ArgumentError: comparison of Fixnum with nil failed
[2] pry(main)> nil < 5
NoMethodError: undefined method `<' for nil:NilClass
```

rrosenblum commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,53 @@
+          RuboCop::AST::Node::COMPARISON_OPERATORS.include?(node.method_name)
+        end
+
+        def yoda_condition?(node)
```
Can this check be written with `NodePattern`?

smakagon replied:
@rrosenblum Thanks for comment. Where I can read more about NodePattern? Do you have some example on top of your mind?

rrosenblum commented on lib/rubocop/cop/style/yoda_condition.rb:
```diff
@@ -0,0 +1,53 @@
+      #   foo == 99
+      #   foo == "bar"
+      class YodaCondition < Cop
+        MSG = 'Reverse the order in conditional statement: `%s` == `%s`.'.freeze
```
The comparison method can be something other than `==`. I think you could change this to `MSG = 'Reverse the order in conditional statement: `%s`.'.freeze`, and pass `node.source` into format.

smakagon replied:
@rrosenblum that's a good point. Thanks.

betesh commented:
> Should this cop be configured to support the opposite style as well?

Yes.  I already commented about this [here](https://github.com/bbatsov/rubocop/issues/4145#issuecomment-287776258)

rrosenblum commented:
I didn't see any tests for other variable types and method calls such as `@foo`, `foo.bar`, `foo[0]`, and `Foo::BAR`.

mikegee commented:
Do we want to allow `1 < x && x < 5` style explicitly, or depend on folks disabling the cop for that style?

jonas054 commented:
@mikegee Good point. I'd like the cop to allow those.

bbatsov commented:
Such things should likely be configurable. And for this particular example - there's also `between?`. :-)

smakagon commented:
@bbatsov you said:
> I think this cop should be extended to cover all comparison operators as well.
5 < x, 5 <= x, etc. Maybe only === should be ignored.

It already covers these cases because I rely on `RuboCop::AST::Node::COMPARISON_OPERATORS` which includes all those operators.

I'm trying to come up with to-do list for this PR.

Seems like it should be the following:
- autocorrection
- configuration (need more info on that, what exactly should be configurable and how can we do that?)
- more cases covered by specs
- @rrosenblum suggested changes the message. what do you think about that?

bbatsov commented:
The plan seems reasonable. I'm ok with the message proposed by @rrosenblum.

bbatsov commented on CHANGELOG.md:
```diff
@@ -1,13 +1,12 @@
 # Change log
 
-## master (unreleased)
```
I guess you removed this accidentally.

smakagon replied:
Yes, that was accidentally. Will return it back.

bbatsov commented:
I still don't see any tests or doc examples about comparison ops that are not `==` (e.g. `>=` and friends).

smakagon commented:
@bbatsov @Drenmi @rrosenblum @mikegee  guys I'm 100% busy on my project right now, can't devote even couple hours to add autocorrection and couple additional specs. If you feel like you can pick up this PR and improve it I'll really appreciate that. At some point I'll definitely get back to it but not in the next couple weeks.

smakagon commented:
Got some free time. Will try to get done low hanging fruits for this PR.

smakagon commented:
Ok, here is current status on this PR:

1. Added autocorrection
2. Changed message as suggested by @rrosenblum 
3. Added more specs (including specs for autocorrection and different variable types)
4. Specs are green on Travis-CI. 
5. Rebased my branch against master.

Let me know what do you thing guys. Didn't work on configuration since I'm not sure what exactly should be configurable.

Thanks.

bbatsov commented:
90% of the tests seems useless to me. You don't really need to check `ifs`, `whens` and so on. On the other hand there are still almost no tests about the other comparison ops. And you haven't updated the cop documentation as well. The code examples you've used are also hard to read. Instead of cramming everything on a single line use a multiline heredoc assigned to a variable as we do pretty much everywhere else.

smakagon commented:
@bbatsov 
Thanks for review.

> 90% of the tests seems useless to me.

Ok, I'll clean up those.


> On the other hand there are still almost no tests about the other comparison ops.

I'll add couple more for `!=` but for other I have specs.


> And you haven't updated the cop documentation as well.

Could you refer to some file or line of description, because it's hard to fix something if I don't understand what and where should be changed :)
I thought I changed description according to all suggestions we have here.

One problem I see that In `config/enabled.yml` current description says:
> 'Do not use constants or literals as the first operand of a comparison'

But I have constant as a whitelisted type, so `FOO == my_var` will not be treated as an offense.

>  The code examples you've used are also hard to read. Instead of cramming everything on a single line use a multiline heredoc assigned to a variable as we do pretty much everywhere else.

Do you mean code examples in specs, right?

smakagon commented:
@bbatsov Removed duplicated specs and added new cases. Let me know if you think there is something I can improve. Thanks.
